### PR TITLE
feat: nft airdrop

### PR DESCRIPTION
**Pull request by zhouhanseng**
closes #

### CODE REVIEW

File: packages/maskbook/src/plugins/ITO/SNSAdaptor/NftAirdropCard.tsx [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    },
+}))
+
+export function NftAirdropCard() {
+    const { t } = useI18N()
+    const classes = useStyles()
+    const title = 'Social Defi Alliance NFT'
+    const endTime = 1628065185808
+    const nfts = new Array(10)
+    nfts.fill({ img: new URL('../assets/nft.png', import.meta.url).toString() })
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/ITO/SNSAdaptor/NftAirdropCard.tsx:94]: Can we get NFT detailed data from the contract?

File: packages/maskbook/src/plugins/ITO/SNSAdaptor/ClaimAllDialog.tsx [hunk 2]
```diff
@@ -322,55 +332,69 @@ export function ClaimAllDialog(props: ClaimAllDialogProps) {
...
-                        </div>
-                    ) : swappedTokens.length > 0 || swappedTokensOld.length > 0 ? (
-                        <>
-                            {swappedTokensOld.length > 0 ? (
-                                <div className={classes.content}>
-                                    {swappedTokens.length > 0 && swappedTokensOld.length > 0 ? (
-                                        <Typography color="textPrimary" className={classes.contentTitle}>
-                                            {t('plugin_ito_claim_all_old_contract')}
+        <SnackbarProvider
+            domRoot={DialogRef.current as HTMLElement}
```
Comment (zhouhanseng) [inline packages/maskbook/src/plugins/ITO/SNSAdaptor/ClaimAllDialog.tsx:336]: https://iamhosseindhv.com/notistack/demos#react-portal

Otherwise, snackbar would be covered under shadow root dialog

File: packages/maskbook/src/plugins/ITO/SNSAdaptor/hooks/useSpaceStationContractClaimCallback.ts [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+            return
+        }
+
+        // start waiting for provider to confirm tx
+        setClaimState({
+            type: TransactionStateType.WAIT_FOR_CONFIRMING,
+        })
+
+        const useSignature = await Services.Ethereum.personalSign(
+            `${campaignInfo.name}
```
Comment (zhouhanseng) [inline packages/maskbook/src/plugins/ITO/SNSAdaptor/hooks/useSpaceStationContractClaimCallback.ts:38]: The signature message is just this format

File: packages/maskbook/src/plugins/ITO/SNSAdaptor/NftAirdropCard.tsx [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+                                    })}
+                                </Typography>
+                            ) : null}
+                        </div>
+                        <div className={classes.claimWrapper}>
+                            <div className={classes.nftsWrapper}>
+                                <div className={classes.gallery}>
+                                    {claimable
+                                        ? campaignInfo.nfts.map((nft) => (
+                                              <div className={classes.imgWrapper}>
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/ITO/SNSAdaptor/NftAirdropCard.tsx:240]: The `key` is lost.

File: packages/maskbook/src/plugins/ITO/SNSAdaptor/NftAirdropCard.tsx [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+        autoHideDuration: 100000,
+        anchorOrigin: {
+            vertical: 'top',
+            horizontal: 'center',
+        },
+    }
+    useEffect(() => setCheckAddress(''), [account])
+
+    useEffect(() => {
+        console.log({ claimState })
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/ITO/SNSAdaptor/NftAirdropCard.tsx:163]:
```suggestion
```

File: packages/maskbook/src/plugins/ITO/SNSAdaptor/hooks/useSpaceStationContractClaimCallback.ts [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+        const { allow, signature, verifyIDs, nftCoreAddress, powahs } = await getAccountClaimSignature(
+            useSignature,
+            account,
+            campaignInfo.chain,
+        )
+
+        if (!allow) {
+            setClaimState({
+                type: TransactionStateType.FAILED,
+                error: new Error('Not allowed to claim'),
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/ITO/SNSAdaptor/hooks/useSpaceStationContractClaimCallback.ts:53]:
```suggestion
                error: new Error('Not allowed to claim.'),
```

File: packages/maskbook/src/plugins/ITO/Worker/apis/spaceStationUUPS.ts [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+        prepareParticipate(input: {
+            signature:"${userSignature}",
+            campaignID: 92,
+            address: "${account}",
+            mintCount: 1,
+            chain: ${chain}
+        }){
+            allow
+            signature
+            mintFuncInfo{
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/ITO/Worker/apis/spaceStationUUPS.ts:130]:
```suggestion
            mintFuncInfo {
```

File: packages/maskbook/src/plugins/ITO/SNSAdaptor/NftAirdropCard.tsx [hunk 8]
```diff
@@ -0,0 +1,315 @@
...
+                                placeholder="Enter your wallet address"
+                            />
+                            <Button
+                                disabled={spaceStationAccountClaimableLoading || checkAddress === ''}
+                                onClick={async () => spaceStationAccountClaimableCallback(checkAddress)}
+                                className={classes.actionButton}>
+                                <span>{t('plugin_airdrop_nft_check')}</span>
+                                {spaceStationAccountClaimableLoading ? (
+                                    <CircularProgress size={16} className={classes.loading} />
+                                ) : null}
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/ITO/SNSAdaptor/NftAirdropCard.tsx:291]:
```suggestion
                                {spaceStationAccountClaimableLoading ? (
                                    <CircularProgress size={16} className={classes.loading} />
                                ) : null}
                                <span>{t('plugin_airdrop_nft_check')}</span>
```

File: packages/maskbook/src/plugins/ITO/SNSAdaptor/NftAirdropCard.tsx [hunk 9]
```diff
@@ -0,0 +1,315 @@
...
+                {
+                    variant: 'success',
+                    ...snackbarOptions,
+                } as OptionsObject,
+            )
+        } else if (claimState.type === TransactionStateType.FAILED) {
+            enqueueSnackbar(
+                <div className={classes.snackbarContent}>
+                    <div className={classes.snackbarTipContent}>
+                        <Typography>{t('plugin_airdrop_nft_claim_all')}</Typography>
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/ITO/SNSAdaptor/NftAirdropCard.tsx:194]:
```suggestion
                        <Typography color="textPrimary">{t('plugin_airdrop_nft_claim_all')}</Typography>
```

File: packages/maskbook/src/plugins/ITO/SNSAdaptor/NftAirdropCard.tsx [hunk 10]
```diff
@@ -0,0 +1,315 @@
...
+    }
+    useEffect(() => setCheckAddress(''), [account])
+
+    useEffect(() => {
+        console.log({ claimState })
+        if (claimState.type === TransactionStateType.CONFIRMED && claimState.no === 0) {
+            enqueueSnackbar(
+                <div className={classes.snackbarContent}>
+                    <div className={classes.snackbarTipContent}>
+                        <Typography>{t('plugin_airdrop_nft_claim_all')}</Typography>
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/ITO/SNSAdaptor/NftAirdropCard.tsx:168]:
```suggestion
                        <Typography color="textPrimary">{t('plugin_airdrop_nft_claim_all')}</Typography>
```

File: packages/maskbook/src/plugins/ITO/SNSAdaptor/NftAirdropCard.tsx [hunk 11]
```diff
@@ -0,0 +1,315 @@
...
+        useSpaceStationClaimableTokenCountCallback()
+    const account = useAccount()
+    const { value: claimable, loading: claimableLoading } = useSpaceStationClaimable(account)
+    const loading = claimableLoading || campaignInfoLoading
+    const [claimState, claimCallback] = useSpaceStationContractClaimCallback(campaignInfo!)
+    const theme = useTheme()
+    const { enqueueSnackbar, closeSnackbar } = useSnackbar()
+    const snackbarOptions = {
+        preventDuplicate: true,
+        autoHideDuration: 100000,
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/ITO/SNSAdaptor/NftAirdropCard.tsx:154]:
```suggestion
        autoHideDuration: 100000, /* 100 seconds */
```

File: packages/web3-constants/data/space-station-uups.json [hunk 12]
```diff
@@ -0,0 +1,28 @@
+{
+    "CONTRACT_ADDRESS": {
+        "Mainnet": "",
+        "Ropsten": "",
+        "Rinkeby": "",
+        "Kovan": "",
+        "Gorli": "",
+        "BSC": "",
+        "BSCT": "",
+        "Matic": "",
```
Comment (guanbinrui) [inline packages/web3-constants/data/space-station-uups.json:10]: When to deploy on Matic Mainnet?

Comment (zhouhanseng) [reply packages/web3-constants/data/space-station-uups.json:10]: after mumbai test pass.

**Comment by wenluomask** [comment]
Checked.

File: packages/maskbook/src/plugins/ITO/SNSAdaptor/ClaimAllDialog.tsx [hunk 13]
```diff
@@ -322,103 +367,109 @@ export function ClaimAllDialog(props: ClaimAllDialogProps) {
...
+                                        '&:hover': {
+                                            backgroundColor: '#1854c4',
+                                        },
+                                        minHeight: 'auto',
+                                        width: '100%',
+                                        fontSize: 18,
+                                        fontWeight: 400,
+                                    }}>
+                                    {(swappedTokens && swappedTokens.length > 0) ||
+                                    (swappedTokensOld && swappedTokensOld.length > 0) ? (
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/ITO/SNSAdaptor/ClaimAllDialog.tsx:434]:
```suggestion
                                    {swappedTokens?.length || swappedTokensOld?.length ? (
```

File: packages/maskbook/src/plugins/ITO/SNSAdaptor/NftAirdropCard.tsx [hunk 14]
```diff
@@ -0,0 +1,337 @@
...
+    const { enqueueSnackbar, closeSnackbar } = useSnackbar()
+    const snackbarOptions = {
+        preventDuplicate: true,
+        anchorOrigin: {
+            vertical: 'top',
+            horizontal: 'center',
+        },
+    }
+
+    useEffect(() => setCheckAddress(''), [account])
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/ITO/SNSAdaptor/NftAirdropCard.tsx:177]: Do you forget to add `chainId`?

```suggestion
    useEffect(() => setCheckAddress(''), [account, chainId])
```

File: packages/maskbook/src/plugins/ITO/SNSAdaptor/NftAirdropCard.tsx [hunk 15]
```diff
@@ -0,0 +1,337 @@
...
+                                      date: formatDateTime(campaignInfo.endTime * 1000, 'yyyy-MM-dd HH:mm'),
+                                  })}
+                        </Typography>
+                    </div>
+                    <div className={classes.claimWrapper}>
+                        <div className={classes.nftsWrapper}>
+                            <div className={classes.gallery}>
+                                {campaignInfo.nfts.map((nft, i) => (
+                                    <div className={classes.imgWrapper} key={i}>
+                                        <img src={nft.image} className={classes.nftImage} />{' '}
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/ITO/SNSAdaptor/NftAirdropCard.tsx:248]: Why `' '` between images?

File: packages/maskbook/src/plugins/ITO/Worker/apis/spaceStationUUPS.ts [hunk 16]
```diff
@@ -0,0 +1,181 @@
...
+                verifyIDs: number[]
+                nftCoreAddress: string
+                powahs: number[]
+            }
+        }
+    }>(`
+      mutation {
+        prepareParticipate(input: {
+            signature:"${userSignature}",
+            campaignID: 92,
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/ITO/Worker/apis/spaceStationUUPS.ts:127]:
```suggestion
            campaignID: ${CAMPAIGN_ID},
```

File: packages/maskbook/src/plugins/ITO/Worker/apis/spaceStationUUPS.ts [hunk 17]
```diff
@@ -0,0 +1,181 @@
...
+    account: string,
+    chain: string,
+    txHash: string,
+    verifyIDs: number[],
+) {
+    const data = await fetchFromSubgraph<{ participate: { participated: boolean } }>(`
+    mutation {
+        participate(input: {
+            signature:"${userSignature}",
+            campaignID: 92,
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/ITO/Worker/apis/spaceStationUUPS.ts:167]:
```suggestion
            campaignID: ${CAMPAIGN_ID},
```

**Comment by Jack-Works** [comment]
hi @zhouhanseng this PR has conflicts with the following files:

```
packages/maskbook/src/_locales/en/messages.json
packages/maskbook/src/_locales/ja/messages.json
packages/maskbook/src/_locales/ko/messages.json
packages/maskbook/src/_locales/zh/messages.json
packages/maskbook/src/components/InjectedComponents/ToolboxHint.tsx
packages/maskbook/src/plugins/ITO/SNSAdaptor/ClaimAllDialog.tsx
packages/web3-shared/src/constants/index.ts
pnpm-lock.yaml
```

File: packages/maskbook/src/web3/UI/EthereumWalletConnectedBoundary.tsx [hunk 18]
```diff
@@ -85,7 +86,12 @@ export function EthereumWalletConnectedBoundary(props: EthereumWalletConnectedBo
     if (!chainIdValid && !offChain)
         return (
             <Grid container>
-                <ActionButton className={classes.button} disabled fullWidth variant="contained" size="large">
+                <ActionButton
```
Comment (Jack-Works) [inline packages/maskbook/src/web3/UI/EthereumWalletConnectedBoundary.tsx:89]: Please avoid using ActionButton. Use the standard Button instead.

File: packages/maskbook/src/plugins/ITO/SNSAdaptor/hooks/useSpaceStationCampaignInfo.ts [hunk 19]
```diff
@@ -0,0 +1,6 @@
+import { useAsyncRetry } from 'react-use'
+import { getCampaignInfo } from '../../Worker/apis/spaceStationUUPS'
+
+export function useSpaceStationCampaignInfo() {
+    return useAsyncRetry(getCampaignInfo, [])
```
Comment (Jack-Works) [inline packages/maskbook/src/plugins/ITO/SNSAdaptor/hooks/useSpaceStationCampaignInfo.ts:5]: Are you sure calling those APIs without using RPC?

File: packages/maskbook/src/plugins/ITO/SNSAdaptor/hooks/useSpaceStationClaimable.ts [hunk 20]
```diff
@@ -0,0 +1,12 @@
+import { getClaimableTokenCount } from '../../Worker/apis/spaceStationUUPS'
+import { useAsyncRetry } from 'react-use'
+
+export function useSpaceStationClaimable(address: string) {
+    return useAsyncRetry(async () => {
+        const data = await getClaimableTokenCount(address)
```
Comment (Jack-Works) [inline packages/maskbook/src/plugins/ITO/SNSAdaptor/hooks/useSpaceStationClaimable.ts:6]: Are you sure calling those APIs without using RPC?

File: packages/maskbook/src/plugins/ITO/SNSAdaptor/hooks/useSpaceStationClaimableTokenCountCallback.ts [hunk 21]
```diff
@@ -0,0 +1,23 @@
...
+import { useCallback, useState } from 'react'
+import type { ClaimableCount } from '../../types'
+
+export function useSpaceStationClaimableTokenCountCallback() {
+    const [spaceStationAccountClaimableLoading, setSpaceStationAccountClaimableLoading] = useState(false)
+    const [spaceStationClaimableCount, setSpaceStationClaimableCount] = useState<ClaimableCount>()
+    const spaceStationAccountClaimableCallback = useCallback(
+        async (address: string) => {
+            setSpaceStationAccountClaimableLoading(true)
+            const data = await getClaimableTokenCount(address)
```
Comment (Jack-Works) [inline packages/maskbook/src/plugins/ITO/SNSAdaptor/hooks/useSpaceStationClaimableTokenCountCallback.ts:11]: Are you sure calling those APIs without using RPC?

File: packages/maskbook/src/plugins/ITO/SNSAdaptor/hooks/useSpaceStationCampaignInfo.ts (hunk 19 above)
Comment (zhouhanseng) [reply packages/maskbook/src/plugins/ITO/SNSAdaptor/hooks/useSpaceStationCampaignInfo.ts:5]: Hi, this API should return same data no matter which chain you are on.

File: packages/maskbook/src/plugins/ITO/SNSAdaptor/hooks/useSpaceStationClaimable.ts (hunk 20 above)
Comment (zhouhanseng) [reply packages/maskbook/src/plugins/ITO/SNSAdaptor/hooks/useSpaceStationClaimable.ts:6]: Yes, it's one graphql api, just use `fetch`.

File: packages/maskbook/src/plugins/ITO/SNSAdaptor/hooks/useSpaceStationClaimableTokenCountCallback.ts (hunk 21 above)
Comment (zhouhanseng) [reply packages/maskbook/src/plugins/ITO/SNSAdaptor/hooks/useSpaceStationClaimableTokenCountCallback.ts:11]: yes

File: packages/maskbook/src/web3/UI/EthereumWalletConnectedBoundary.tsx (hunk 18 above)
Comment (zhouhanseng) [reply packages/maskbook/src/web3/UI/EthereumWalletConnectedBoundary.tsx:89]: OK, we will replace `ActionButton` in the future, but not in this PR.

File: packages/maskbook/src/plugins/ITO/SNSAdaptor/ClaimAllDialog.tsx (hunk 13 above)
Comment (zhouhanseng) [reply packages/maskbook/src/plugins/ITO/SNSAdaptor/ClaimAllDialog.tsx:436]: Object is possibly 'undefined'.ts(2532)
const swappedTokens: SwappedToken[] | undefined

**Review by guanbinrui — DISMISSED** [review]

**Comment by Jack-Works** [comment]
Please wait. I'll review tonight

File: packages/maskbook/src/plugins/ITO/SNSAdaptor/hooks/useSpaceStationCampaignInfo.ts (hunk 19 above)
Comment (Jack-Works) [reply packages/maskbook/src/plugins/ITO/SNSAdaptor/hooks/useSpaceStationCampaignInfo.ts:5]: I don't think so. Calling them directly in SNS might cause CSP limitation.

Comment (zhouhanseng) [reply packages/maskbook/src/plugins/ITO/SNSAdaptor/hooks/useSpaceStationCampaignInfo.ts:5]: The API provider is our project cooperator. We are at the same Wechat group.

Comment (Jack-Works) [reply packages/maskbook/src/plugins/ITO/SNSAdaptor/hooks/useSpaceStationCampaignInfo.ts:5]: The CSP limitation is set by the SNS network, not the API provider.

File: .prettierrc [hunk 22]
```diff
@@ -2,6 +2,5 @@
   "trailingComma": "all",
   "printWidth": 120,
   "semi": false,
-  "singleQuote": true,
-  "jsxBracketSameLine": true
```
Comment (septs) [inline .prettierrc]: @Jack-Works ?

Comment (Jack-Works) [reply .prettierrc]: oops that an error

Comment (Jack-Works) [reply .prettierrc:6]: please revert this change. (see develop branch for the latest value)

File: packages/scripts/src/utils/prettier.ts [hunk 23]
```diff
@@ -6,7 +6,6 @@ export async function prettier(f: string) {
         printWidth: 120,
         semi: false,
         singleQuote: true,
-        jsxBracketSameLine: true,
```
Comment (Jack-Works) [inline packages/scripts/src/utils/prettier.ts]: Please revert this change

File: .prettierrc (hunk 22 above)
Comment (zhouhanseng) [reply .prettierrc]: ok

File: packages/maskbook/src/locales/ko-KR.json [hunk 24]
```diff
@@ -267,6 +267,21 @@
...
     "wallet_add_nft_id_placeholder": "수집품 아이디 입력",
+    "wallet_airdrop_nft_unclaimed_title": "",
+    "plugin_airdrop_nft_start_time": "",
+    "plugin_airdrop_nft_end_time": "",
+    "plugin_airdrop_nft_expired": "",
+    "plugin_airdrop_nft_claim": "",
+    "plugin_airdrop_nft_claimed": "",
+    "plugin_airdrop_nft_check": "",
+    "plugin_airdrop_nft_check_address": "",
+    "plugin_airdrop_nft_incorrect_address": "",
```
Comment (Jack-Works) [inline packages/maskbook/src/locales/ko-KR.json:278]: You don't need to add those empty lines

Comment (zhouhanseng) [reply packages/maskbook/src/locales/ko-KR.json:278]: okay, get it!

**Review by guanbinrui — DISMISSED** [review]

**Comment by AinuX** [comment]
Checked

**Review by guanbinrui — APPROVED** [review]

**Review by Jack-Works — APPROVED** [review]
